consensus/vbft/chain_store.go

```diff
@@ -78,58 +139,59 @@ func (self *ChainStore) AddBlock(block *Block) error {
 	if block.Block.Header == nil {
 		panic("nil block header")
 	}
-	self.pendingBlocks[block.getBlockNum()] = block
-
 	blkNum := self.GetChainedBlockNum() + 1
 	for {
-		if blk, present := self.pendingBlocks[blkNum]; blk != nil && present {
-			log.Infof("ledger adding chained block (%d, %d)", blkNum, self.GetChainedBlockNum())
-
-			err := self.db.AddBlock(blk.Block)
-			if err != nil && blkNum > self.GetChainedBlockNum() {
-				return fmt.Errorf("ledger add blk (%d, %d) failed: %s", blkNum, self.GetChainedBlockNum(), err)
-			}
-
-			self.chainedBlockNum = blkNum
-			if blkNum != self.db.GetCurrentBlockHeight() {
-				log.Errorf("!!! chain store added chained block (%d, %d): %s",
-					blkNum, self.db.GetCurrentBlockHeight(), err)
+		var err error
+		if self.needSubmitBlock {
+			if submitBlk, present := self.pendingBlocks[blkNum-1]; submitBlk != nil && present {
+				err := self.db.SubmitBlock(submitBlk.block.Block, *submitBlk.execResult)
+				if err != nil && blkNum > self.GetChainedBlockNum() {
+					return fmt.Errorf("ledger add submitBlk (%d, %d) failed: %s", blkNum, self.GetChainedBlockNum(), err)
+				}
+				if _, present := self.pendingBlocks[blkNum-2]; present {
+					delete(self.pendingBlocks, blkNum-2)
+				}
+			} else {
+				break
 			}
-
-			delete(self.pendingBlocks, blkNum)
-			blkNum++
-		} else {
-			break
 		}
+		execResult, err := self.db.ExecuteBlock(block.Block)
+		if err != nil {
+			log.Errorf("chainstore AddBlock GetBlockExecResult: %s", err)
+			return fmt.Errorf("chainstore AddBlock GetBlockExecResult: %s", err)
+		}
+		self.pendingBlocks[blkNum] = &PendingBlock{block: block, execResult: &execResult}
+		self.needSubmitBlock = true
+		self.server.pid.Tell(
+			&message.BlockConsensusComplete{
+				Block: block.Block,
+			})
+		self.chainedBlockNum = blkNum
+		blkNum++
+		break
 	}
-
 	return nil
 }
 
-//
-// SetBlock is used when recovering from fork-chain
-//
-func (self *ChainStore) SetBlock(block *Block, blockHash common.Uint256) error {
-
-	err := self.db.AddBlock(block.Block)
-	self.chainedBlockNum = self.db.GetCurrentBlockHeight()
-	if err != nil {
-		return fmt.Errorf("ledger failed to add block: %s", err)
-	}
-
-	return nil
+func (self *ChainStore) SetBlock(blkNum uint32, blk *PendingBlock) {
```

Comment:
It seems SetBlock() only used when opening chain store.

Reply:
has delete